aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1344 @@
+
+    /// Interface to manage the transaction execution queue.
+    ///
+    /// Allows adding or removing transactions from the execution queue based on
+    /// their status changes. Used when stalls are added/removed or when
+    /// a new incarnation is created.
+    execution_queue_manager: &'a ExecutionQueueManager,
+
+    /// Index of this transaction in the block.
+    txn_idx: TxnIndex,
+}
+
+impl<'a> ExecutionStatus<'a> {
+    pub(crate) fn new(
+        execution_queue_manager: &'a ExecutionQueueManager,
+        txn_idx: TxnIndex,
+    ) -> Self {
+        Self {
+            inner_status: CachePadded::new(Mutex::new(InnerStatus::new())),
+            next_incarnation_to_abort: CachePadded::new(AtomicU32::new(0)),
+            dependency_shortcut: CachePadded::new(AtomicU8::new(DependencyFlag::Defer as u8)),
+            num_stalls: CachePadded::new(AtomicU32::new(0)),
+            execution_queue_manager,
+            txn_idx,
+        }
+    }
+
+    /// Attempts to transition a transaction from PendingScheduling to Executing state.
+    ///
+    /// This method is called by the scheduler when it selects a transaction for execution.
+    /// It only succeeds if the transaction is currently in PendingScheduling state.
+    ///
+    /// # Returns
+    /// - `Ok(Some(incarnation))` if the transition was successful, returning the incarnation number
+    /// - `Ok(None)` if the transaction is not in PendingScheduling state
+    /// - `Err(PanicError)` if there was an error during flag transition
+    pub(crate) fn try_start_executing(&self) -> Result<Option<Incarnation>, PanicError> {
+        let inner_status = &mut *self.inner_status.lock();
+        let ret = inner_status.try_start_executing();
+
+        if ret.is_some() {
+            // When status is PendingScheduling the dependency shortcut flag should be
+            // DEFER (default or set by abort under the inner status lock).
+            self.swap_dependency_flag_any(
+                &[DependencyFlag::Defer],
+                DependencyFlag::Executing,
+                "try_start_executing",
+            )?;
+        }
+
+        Ok(ret)
+    }
+
+    /// Marks a transaction's execution as completed.
+    ///
+    /// Called when a transaction finishes executing, whether successfully or after being aborted.
+    /// Updates the transaction's status based on its current state:
+    /// - If Executing → Executed (successful execution)
+    /// - If Aborted → PendingScheduling with incremented incarnation (ready for re-execution)
+    ///
+    /// # Parameters
+    /// - `finished_incarnation`: The incarnation that has finished execution
+    ///
+    /// # Returns
+    /// - 'Ok(is the execution still valid at the time finish was recorded)', in particular:
+    ///     - `Ok(true)` if transitioned from Executing to Executed
+    ///     - `Ok(false)` if transitioned from Aborted to PendingScheduling
+    /// - `Err` if the current state doesn't allow finishing execution
```

Comment:
I think one thing that can help reviewers is to document what can be running in parallel and what can only be run by one thread e.g. I'm assuming only one thread can take an incarnation of a txn but multiple threads can try to abort one txn. maybe the naming prefix `try_` is an indicator of it but being more explicit is helpful

Reply:
Let me add some clarifications about these to the most important methods.

Reply:
In general, I think everything can pretty much be concurrent except:
- add_stall must finish for the corresponding remove_stall, but in the meantime there might even be other concurrent add and remove stalls on the same status.
- multiple tries might actually be called, but only one succeeding means there is only a single finish execution (and can only be concurrent with abort calls - theoretically with outdated incarnation calls as well but those are kind of void).

Now I think maybe I should just write the above in the top level comment?

Reply:
I don't think finish_execution, finish_abort is concurrent or do I miss something?

Reply:
Unless I forgot myself, which would be funny, I think they can actually be concurrent. There can only be 1 finish_execution and 1 finish_abort though. Finish abort would be the "winner" of start_aborts and finish execution would be the winner of "start_executing" (renamed try_start_executing). If finish execution goes first, then update would be to "Executed", and abort will change it to pending, while if finish abort goes first, it would go to "aborted" (I changed this since you had a look at the early draft to make things more symmetrical and explicit) and then to pending. Diagram is also updated at the file level comments.

Reply:
sorry by concurrent I mean multiple call of the "same" function